### ISSUE TITLE
DuckDNS API call incorrect

### ISSUE THREAD

**Issue by jimp100** [acmesh-official/acme.sh#3285]
dns_duckdns integration makes an incorrect API call.

Steps to reproduce
------------------
Make a acme.sh call for DuckDNS.  
```
acme.sh --insecure --issue --dns dns_duckdns -d '*.mydomain.duckdns.org'
```
Note, this isn't isolated to wildcard certs, issue occurs for mydomain.duckdns.org

Debug log
-----------------
```
[Sat Dec  5 13:43:45 GMT 2020] Trying to add TXT record
[Sat Dec  5 13:43:45 GMT 2020] param='domains=_acme-challenge.xxxxxx.duckdns.org&token=aaaaaaaa-bbbb-cccc-dddd-eeeeeeeeeeee&txt=b2k7MoBJ89KWIjAXwg9XjjJcDs_4B8o_7YXDWddoQJI'
[Sat Dec  5 13:43:45 GMT 2020] url='https://www.duckdns.org/update?domains=_acme-challenge.xxxxxx.duckdns.org&token=aaaaaaaa-bbbb-cccc-dddd-eeeeeeeeeeee&txt=b2k7MoBJ89KWIjAXwg9XjjJcDs_4B8o_7YXDWddoQJI&verbose=true'
[Sat Dec  5 13:43:45 GMT 2020] GET
[Sat Dec  5 13:43:45 GMT 2020] url='https://www.duckdns.org/update?domains=_acme-challenge.xxxxxx.duckdns.org&token=aaaaaaaa-bbbb-cccc-dddd-eeeeeeeeeeee&txt=b2k7MoBJ89KWIjAXwg9XjjJcDs_4B8o_7YXDWddoQJI&verbose=true'
[Sat Dec  5 13:43:45 GMT 2020] timeout=
[Sat Dec  5 13:43:45 GMT 2020] _CURL='curl -L --silent --dump-header /root/.acme.sh/http.header  -g  --insecure  '
[Sat Dec  5 13:43:46 GMT 2020] ret='0'
[Sat Dec  5 13:43:46 GMT 2020] Errors happened during adding the TXT record, response=KO
```
Notice the generated request URL is as follows:
`
https://www.duckdns.org/update?domains=_acme-challenge.mydomain.duckdns.org&token=aaaa......
`

The correct request URL should not include the `_acme-challenge.` subdomain
`
https://www.duckdns.org/update?domains=mydomain.duckdns.org&token=aaaa......
`

DuckDNS documentation: https://www.duckdns.org/spec.jsp

**Comment by nate1010smith** [comment]
I just encountered this on a freebsd host running acme.sh version 2.8.7.

As far as I can tell the issue is that POSIX Basic Regular Expressions don't support '?' for groups. The sed command that extracts the duckdns domain uses that. At least on freebsd, sed doesn't support that which causes '_acme-challenge' to get left in the domain that goes into the API request.

Replacing '\?' with '\{0,1\}' retains the intended functionality while being a proper POSIX BRE. That fixes the bug when using BSD sed and should function equivalently when using GNU sed.

**Comment by jimp100** [comment]
I too am using acme.sh 2.8.7 on freebsd.  From my test, your proposed change causes just the subdomain (i.e. **mydomain**) to be returned, the primary domain **duckdns.org** is being removed.

**Comment by nate1010smith** [comment]
@jimp100, I believe that's the  intended  behavior for the line I changed per the comments above the `_duckdns_get_domain()` function. That agrees with the DuckDNS API spec which says "The domain does not need to include the .duckdns.org part of your domain, just the subname."

That change worked for me and allowed me to refresh my cert.

**Comment by jimp100** [comment]
@nate1010smith ah yes sorry, you are correct.  The API accepts both with and without the primary domain, which has thrown me.  Thanks

**Comment by jimp100** [comment]
@nate1010smith I've subsequently noticed the same line fails to get the duckdns domain when requesting a certificate for a specific subdomain of it. For example: `mysubdomain.mydomain.duckdns.org`

**Comment by nate1010smith** [comment]
@jimp100, I think you're correct that the current code fails for sub-subdomains. I'm not using any sub-subdomains and don't have an environment set up for testing so I don't plan to submit a patch.

By my reading of the Duck DNS API spec, I think the correct behavior for `subsubdomain.mydomain.duckdns.org` would be to update the TXT record for `mydomain`  which would then apply to `subsubdomain` as well.

I think something like this should do it:
```sh
_duckdns_domain="$(printf "%s" "$fulldomain" | _lower_case | sed -n 's/^\([^.]\{1,\}\.\)*\([a-z0-9-]\{1,\}\)\.duckdns\.org$/\2/p;')"
```

If that works for you, I'd suggest submitting a patch or coordinating with @raidenii, who appears be the maintainer for the Duck DNS module according to https://github.com/acmesh-official/acme.sh/wiki/dnsapi

**Comment by raidenii** [comment]
Thanks @nate1010smith  for the remainder and I really haven't touched this for a LONG while. More than happy to transfer the ownership to whoever wants to maintain it. Meanwhile, I'll test the script to support wildcard and submit a PR for that.

**Comment by DevPump** [comment]
> @jimp100, I think you're correct that the current code fails for sub-subdomains. I'm not using any sub-subdomains and don't have an environment set up for testing so I don't plan to submit a patch.
> 
> By my reading of the Duck DNS API spec, I think the correct behavior for `subsubdomain.mydomain.duckdns.org` would be to update the TXT record for `mydomain` which would then apply to `subsubdomain` as well.
> 
> I think something like this should do it:
> 
> ```shell
> _duckdns_domain="$(printf "%s" "$fulldomain" | _lower_case | sed -n 's/^\([^.]\{1,\}\.\)*\([a-z0-9-]\{1,\}\)\.duckdns\.org$/\2/p;')"
> ```
> 
> If that works for you, I'd suggest submitting a patch or coordinating with @raidenii, who appears be the maintainer for the Duck DNS module according to https://github.com/acmesh-official/acme.sh/wiki/dnsapi



This exactly resolves the problem for me with duckdns wildcard certs.

**Comment by figurcoe** [comment]
Hi, I stumbled across a similar issue. I adjusted the regex for myself and it worked flawlessly for me. Feel free to check if it suits you, see #3340

**Comment by jimp100** [comment]
@figurcoe I believe the issue has been resolved under #3307 which has been merged, I'm not sure if another modification is necessary.  Regardless, I will close this issue as the initial issue is no longer relevant.
